design/Extend-VolumePolicies-to-support-more-actions.md

```diff
@@ -0,0 +1,304 @@
+	return false, nil
+}
+
+// GetVolumesMatchingFSBackupAction returns a list of volume names to backup for the provided pod having fs-backup volume policy action
+func GetVolumesMatchingFSBackupAction(pod *corev1api.Pod, volumePolicy *resourcepolicies.Policies) ([]string, []string, error) {
+	ActionMatchingVols := []string{}
+	NonActionMatchingVols := []string{}
+	for _, vol := range pod.Spec.Volumes {
+		action, err := volumePolicy.GetMatchAction(vol)
+		if err != nil {
+			return nil, nil, err
+		}
+		// Now if the matched action is `fs-backup` then add that Volume to the fsBackupVolumeList
+		if action != nil && action.Type == resourcepolicies.FSBackup {
+			ActionMatchingVols = append(ActionMatchingVols, vol.Name)
+		} else {
+			NonActionMatchingVols = append(NonActionMatchingVols, vol.Name)
+		}
+	}
+
+	return ActionMatchingVols, NonActionMatchingVols, nil
+}
+```
+- The main function from the above `vph.ProcessVolumePolicyFSbackup` will be called when we encounter Pods during the backup workflow:
+```go
+includedVolumes, optedOutVolumes, err := vph.ProcessVolumePolicyFSBackup(pod, vph.VolumePolicy, vph.DefaultVolumesToFsBackup, vph.BackupExcludePVC)
+			if err != nil {
+				backupErrs = append(backupErrs, errors.WithStack(err))
+			}
+```
+#### Snapshot (PV)
+
+- Making sure that `snapshot` action is skipped for PVs that do not fit the volume policy criteria, for this we will use the `vph.ShouldPerformSnapshot` from the `VolumePolicyHelper(vph)` receiver.
+```go
+func (v *VolumePolicyHelper) ShouldPerformSnapshot(obj runtime.Unstructured) (bool, error) {
+	// Here check if the object(pv/pvc) fits a volume policy criteria and see if the associated action is snapshot
+	// if it is not snapshot then skip the code path for snapshotting the PV/PVC
+	action, err := v.VolumePolicy.GetMatchAction(obj)
+	if err != nil {
+		return false, err
+	}
+
+	if action != nil && action.Type == resourcepolicies.Snapshot {
+		return true, nil
+	}
+
+	return false, nil
+
+}
+```
+
+- The above function will be used as follows in `takePVSnapshot` function of the backup workflow:
+```go
+snapshotVolumePolicy, err := vhp.ShouldPerformSnapshot(obj)
+
+	if err != nil {
+		return err
+	}
+
+	if !snapshotVolumePolicy {
+		log.Info(fmt.Sprintf("skipping volume snapshot for PV %s as it does not fit the volume policy criteria for snapshot action", pv.Name))
+		ib.trackSkippedPV(obj, kuberesource.PersistentVolumes, volumeSnapshotApproach, "does not satisfy the criteria for volume policy based snapshot action", log)
+		return nil
+	}
+```
+#### Snapshot (PVC)
+
+- Making sure that `snapshot` action is skipped for PVCs that do not fit the volume policy criteria, for this we will again use the `vph.ShouldPerformSnapshot` from the `VolumePolicyHelper(vph)` receiver.
+- We will pass the `VolumePolicyHelper(vph)` instance in `executeActions` method so that its available to use.
+```go
+func (v *VolumePolicyHelper) ShouldPerformSnapshot(obj runtime.Unstructured) (bool, error) {
```

Comment:
This function should also take other flags, like "snapshotVolumes" in the backup CR.

Reply:
Done, updated the PR